OSGi: ClassNotFoundException: javax.security.cert.X509Certificate cannot be found by io.etcd.jetcd-all_0.3.0.SNAPSHOT

I'm hitting this while playing over in https://github.com/vorburger/opendaylight-etcd :

```
[INFO] --- maven-surefire-plugin:2.18.1:test (default) @ odl-etcd-demo-restconf ---
[INFO] Surefire report directory: /home/vorburger/dev/ODL/git.etcd/opendaylight-etcd/features/odl-etcd-demo-restconf/target/surefire-reports

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.opendaylight.odlparent.featuretest.SingleFeatureTest

Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 41.393 sec <<< FAILURE! - in org.opendaylight.odlparent.featuretest.SingleFeatureTest
installFeatureCatchAndLog(org.opendaylight.odlparent.featuretest.SingleFeatureTest)[repoUrl: file:/home/vorburger/dev/ODL/git.etcd/opendaylight-etcd/features/odl-etcd-demo-restconf/target/feature/feature.xml, Feature: odl-etcd-demo-restconf 0.0.1.SNAPSHOT]  Time elapsed: 39.831 sec  <<< ERROR!
org.opendaylight.odlparent.bundlestest.lib.SystemStateFailureException: diag failed; some bundles failed to start
diag: Failure {Installed=0, Resolved=4, Unknown=0, GracePeriod=9, Waiting=0, Starting=0, Active=295, Stopping=0, Failure=1}

(...)

7. NOK org.opendaylight.etcd.ds:0.0.1.SNAPSHOT: OSGi state = Active, Karaf bundleState = Failure, due to: Blueprint
6/9/18 5:21 PM
Exception: 
org.osgi.service.blueprint.container.ComponentDefinitionException: Error when instantiating bean etcdDOMDataBroker of class org.opendaylight.etcd.ds.impl.EtcdDOMDataBroker
org.osgi.service.blueprint.container.ComponentDefinitionException: org.osgi.service.blueprint.container.ComponentDefinitionException: Error when instantiating bean etcdDOMDataBroker of class org.opendaylight.etcd.ds.impl.EtcdDOMDataBroker
	at org.apache.aries.blueprint.container.ServiceRecipe.createService(ServiceRecipe.java:310)
	at org.apache.aries.blueprint.container.ServiceRecipe.internalGetService(ServiceRecipe.java:252)
	at org.apache.aries.blueprint.container.ServiceRecipe.internalCreate(ServiceRecipe.java:149)
	at org.apache.aries.blueprint.di.AbstractRecipe$1.call(AbstractRecipe.java:79)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.apache.aries.blueprint.di.AbstractRecipe.create(AbstractRecipe.java:88)
	at org.apache.aries.blueprint.container.BlueprintRepository.createInstances(BlueprintRepository.java:255)
	at org.apache.aries.blueprint.container.BlueprintRepository.createAll(BlueprintRepository.java:186)
	at org.apache.aries.blueprint.container.BlueprintContainerImpl.instantiateEagerComponents(BlueprintContainerImpl.java:704)
	at org.apache.aries.blueprint.container.BlueprintContainerImpl.doRun(BlueprintContainerImpl.java:410)
	at org.apache.aries.blueprint.container.BlueprintContainerImpl.run(BlueprintContainerImpl.java:275)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.apache.aries.blueprint.container.ExecutorServiceWrapper.run(ExecutorServiceWrapper.java:106)
	at org.apache.aries.blueprint.utils.threading.impl.DiscardableRunnable.run(DiscardableRunnable.java:48)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.osgi.service.blueprint.container.ComponentDefinitionException: Error when instantiating bean etcdDOMDataBroker of class org.opendaylight.etcd.ds.impl.EtcdDOMDataBroker
	at org.apache.aries.blueprint.container.BeanRecipe.wrapAsCompDefEx(BeanRecipe.java:361)
	at org.apache.aries.blueprint.container.BeanRecipe.getInstanceFromType(BeanRecipe.java:351)
	at org.apache.aries.blueprint.container.BeanRecipe.getInstance(BeanRecipe.java:282)
	at org.apache.aries.blueprint.container.BeanRecipe.internalCreate2(BeanRecipe.java:830)
	at org.apache.aries.blueprint.container.BeanRecipe.internalCreate(BeanRecipe.java:811)
	at org.apache.aries.blueprint.di.AbstractRecipe$1.call(AbstractRecipe.java:79)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.apache.aries.blueprint.di.AbstractRecipe.create(AbstractRecipe.java:88)
	at org.apache.aries.blueprint.di.RefRecipe.internalCreate(RefRecipe.java:62)
	at org.apache.aries.blueprint.di.AbstractRecipe.create(AbstractRecipe.java:106)
	at org.apache.aries.blueprint.container.ServiceRecipe.createService(ServiceRecipe.java:285)
	... 21 more
Caused by: org.opendaylight.etcd.ds.impl.EtcdException: failed to connect (in time) to etcd server
	at org.opendaylight.etcd.ds.impl.EtcdServerUtils.getServerHeader(EtcdServerUtils.java:40)
	at org.opendaylight.etcd.ds.impl.EtcdServerUtils.getServerRevision(EtcdServerUtils.java:33)
	at org.opendaylight.etcd.ds.impl.EtcdDOMDataBrokerProvider.init(EtcdDOMDataBrokerProvider.java:90)
	at org.opendaylight.etcd.ds.impl.EtcdDOMDataBroker.<init>(EtcdDOMDataBroker.java:43)
	at org.opendaylight.etcd.ds.impl.EtcdDOMDataBroker.<init>(EtcdDOMDataBroker.java:36)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.apache.aries.blueprint.utils.ReflectionUtils.newInstance(ReflectionUtils.java:331)
	at org.apache.aries.blueprint.container.BeanRecipe.newInstance(BeanRecipe.java:984)
	at org.apache.aries.blueprint.container.BeanRecipe.getInstanceFromType(BeanRecipe.java:349)
	... 30 more
Caused by: java.util.concurrent.ExecutionException: java.util.concurrent.ExecutionException: io.etcd.jetcd.shaded.io.grpc.StatusRuntimeException: INTERNAL: Panic! This is a bug!
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
	at org.opendaylight.etcd.ds.impl.EtcdServerUtils.getServerHeader(EtcdServerUtils.java:38)
	... 41 more
Caused by: java.util.concurrent.ExecutionException: io.etcd.jetcd.shaded.io.grpc.StatusRuntimeException: INTERNAL: Panic! This is a bug!
	at io.etcd.jetcd.shaded.com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:500)
	at io.etcd.jetcd.shaded.com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:479)
	at io.etcd.jetcd.internal.impl.Util.lambda$toCompletableFutureWithRetry$1(Util.java:128)
	... 3 more
Caused by: io.etcd.jetcd.shaded.io.grpc.StatusRuntimeException: INTERNAL: Panic! This is a bug!
	at io.etcd.jetcd.shaded.io.grpc.Status.asRuntimeException(Status.java:526)
	at io.etcd.jetcd.shaded.io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:468)
	at io.etcd.jetcd.shaded.io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.etcd.jetcd.shaded.io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.etcd.jetcd.shaded.io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at io.etcd.jetcd.internal.impl.ClientConnectionManager$AuthTokenInterceptor$1$1.onClose(ClientConnectionManager.java:308)
	at io.etcd.jetcd.shaded.io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.etcd.jetcd.shaded.io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.etcd.jetcd.shaded.io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at io.etcd.jetcd.shaded.io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:684)
	at io.etcd.jetcd.shaded.io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.etcd.jetcd.shaded.io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.etcd.jetcd.shaded.io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at io.etcd.jetcd.shaded.io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:403)
	at io.etcd.jetcd.shaded.io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:459)
	at io.etcd.jetcd.shaded.io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63)
	at io.etcd.jetcd.shaded.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:546)
	at io.etcd.jetcd.shaded.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:467)
	at io.etcd.jetcd.shaded.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:584)
	at io.etcd.jetcd.shaded.io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.etcd.jetcd.shaded.io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	... 3 more
Caused by: java.lang.NoClassDefFoundError: javax/security/cert/X509Certificate
	at io.etcd.jetcd.shaded.io.netty.util.internal.EmptyArrays.<clinit>(EmptyArrays.java:38)
	at io.etcd.jetcd.shaded.io.netty.util.ResourceLeakDetector.<clinit>(ResourceLeakDetector.java:525)
	at io.etcd.jetcd.shaded.io.netty.buffer.AbstractByteBufAllocator.<clinit>(AbstractByteBufAllocator.java:34)
	at io.etcd.jetcd.shaded.io.netty.buffer.Unpooled.<clinit>(Unpooled.java:74)
	at io.etcd.jetcd.shaded.io.netty.handler.codec.http2.Http2ConnectionHandler.<clinit>(Http2ConnectionHandler.java:74)
	at io.etcd.jetcd.shaded.io.grpc.netty.NettyClientTransport.start(NettyClientTransport.java:192)
	at io.etcd.jetcd.shaded.io.grpc.internal.ForwardingConnectionClientTransport.start(ForwardingConnectionClientTransport.java:32)
	at io.etcd.jetcd.shaded.io.grpc.internal.ForwardingConnectionClientTransport.start(ForwardingConnectionClientTransport.java:32)
	at io.etcd.jetcd.shaded.io.grpc.internal.InternalSubchannel.startNewTransport(InternalSubchannel.java:246)
	at io.etcd.jetcd.shaded.io.grpc.internal.InternalSubchannel.obtainActiveTransport(InternalSubchannel.java:206)
	at io.etcd.jetcd.shaded.io.grpc.internal.ManagedChannelImpl$SubchannelImpl.requestConnection(ManagedChannelImpl.java:1417)
	at io.etcd.jetcd.shaded.io.grpc.PickFirstBalancerFactory$PickFirstBalancer.handleResolvedAddressGroups(PickFirstBalancerFactory.java:74)
	at io.etcd.jetcd.shaded.io.grpc.internal.AutoConfiguredLoadBalancerFactory$AutoConfiguredLoadBalancer.handleResolvedAddressGroups(AutoConfiguredLoadBalancerFactory.java:106)
	at io.etcd.jetcd.shaded.io.grpc.internal.ManagedChannelImpl$NameResolverListenerImpl$1NamesResolved.run(ManagedChannelImpl.java:1286)
	at io.etcd.jetcd.shaded.io.grpc.internal.ChannelExecutor.drain(ChannelExecutor.java:73)
	at io.etcd.jetcd.shaded.io.grpc.internal.ManagedChannelImpl$LbHelperImpl.runSerialized(ManagedChannelImpl.java:1230)
	at io.etcd.jetcd.shaded.io.grpc.internal.ManagedChannelImpl$NameResolverListenerImpl.onAddresses(ManagedChannelImpl.java:1290)
	at io.etcd.jetcd.resolver.SmartNameResolver.doResolve(SmartNameResolver.java:147)
	... 3 more
Caused by: java.lang.ClassNotFoundException: javax.security.cert.X509Certificate cannot be found by io.etcd.jetcd-all_0.3.0.SNAPSHOT
	at org.eclipse.osgi.internal.loader.BundleLoader.findClassInternal(BundleLoader.java:484)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:395)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:387)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:150)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 21 more
```

I guess the `MANIFEST.MF` of the jetcd-all-*.jar is missing `javax.security.cert` in its `Import-Package` ..

@lburgazzoli does jetcd-all-*.jar work for you under OSGi?

--- a/jetcd-all/pom.xml
+++ b/jetcd-all/pom.xml
@@ -127,6 +127,7 @@
                         </Export-Package>
                         <Import-Package>
                             org.slf4j;version="[1.7,2)",
+                            javax.security.cert,
                             javax.net.ssl;resolution:=optional
                         </Import-Package>
                     </instructions>

--- a/jetcd-launcher/pom.xml
+++ b/jetcd-launcher/pom.xml
@@ -28,7 +28,6 @@
     </parent>
 
     <artifactId>jetcd-launcher</artifactId>
-    <packaging>bundle</packaging>
 
     <properties>
         <checkstyle.config.location>${basedir}/../etc/checkstyle.xml</checkstyle.config.location>
@@ -63,17 +62,6 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-checkstyle-plugin</artifactId>
-            </plugin>
-            <plugin>
-                <groupId>org.apache.felix</groupId>
-                <artifactId>maven-bundle-plugin</artifactId>
-                <extensions>true</extensions>
-                <configuration>
-                    <instructions>
-                        <Bundle-SymbolicName>${project.groupId}.${project.artifactId}</Bundle-SymbolicName>
-                        <Bundle-Name>etcd :: ${project.artifactId}</Bundle-Name>
-                    </instructions>
-                </configuration>
             </plugin>
         </plugins>
     </build>

--- a/jetcd-osgi/jetcd-karaf/pom.xml
+++ b/jetcd-osgi/jetcd-karaf/pom.xml
@@ -108,6 +108,11 @@
             <artifactId>assertj-core</artifactId>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>jetcd-launcher</artifactId>
+        </dependency>
     </dependencies>
 
     <build>

--- a/jetcd-osgi/jetcd-karaf/src/test/java/io/etcd/jetcd/osgi/ClientServiceChecks.java
+++ b/jetcd-osgi/jetcd-karaf/src/test/java/io/etcd/jetcd/osgi/ClientServiceChecks.java
@@ -30,8 +30,11 @@ import static org.ops4j.pax.exam.karaf.options.KarafDistributionOption.keepRunti
 import static org.ops4j.pax.exam.karaf.options.KarafDistributionOption.logLevel;
 
 import io.etcd.jetcd.Client;
+import io.etcd.jetcd.data.ByteSequence;
 import io.etcd.jetcd.resolver.URIResolver;
 import java.io.File;
+import java.nio.charset.Charset;
+import java.util.concurrent.TimeUnit;
 import javax.inject.Inject;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -51,7 +54,8 @@ import org.osgi.framework.Constants;
 
 @RunWith(PaxExam.class)
 @ExamReactorStrategy(PerClass.class)
-public class ClientServiceTest extends TestSupport {
+public class ClientServiceChecks extends TestSupport {
+
   @Inject
   protected BundleContext bundleContext;
 
@@ -109,7 +113,7 @@ public class ClientServiceTest extends TestSupport {
             .artifactId("assertj-core")
             .versionAsInProject()
             .start(),
-        editConfigurationFilePut("etc/io.etcd.jetcd.cfg", "endpoints", "localhost:2379"),
+        editConfigurationFilePut("etc/io.etcd.jetcd.cfg", "endpoints", PaxExamWrapperTest.getClientEndpoints()),
         editConfigurationFilePut("etc/io.etcd.jetcd.resolver.dnssrv.cfg", "foo", "bar"),
         keepRuntimeFolder(),
         cleanCaches(),
@@ -118,10 +122,23 @@ public class ClientServiceTest extends TestSupport {
   }
 
   @Test
-  public void testServiceAvailability() {
+  public void testServiceAvailability() throws Exception {
     assertThat(bundleContext).isNotNull();
     assertThat(client).isNotNull();
     assertThat(uriResolver).isNotNull();
-  }
 
+    try {
+        client.getKVClient().get(ByteSequence.from("non-existing", Charset.forName("UTF-8"))).get(13, TimeUnit.SECONDS);
+    } catch (Throwable t) {
+        // Pax Exam's WrappedTestContainerException unfortunately only includes the message, not the cause,
+        // so the real reason for failures needs to be searched for in target/exam/*/data/log/karaf.log ...
+        // Just for convenience in local debugging, and to eastily understand failures on CI, we dump
+        // the failure's stack trace to STDOUT; like that it's easy to see in maven-surefire-plugin,
+        // or when running in the IDE, like for any other non-OSGi test failure.
+        t.printStackTrace();
+        throw t;
+    }
+    // NB: Any NoClassDefFoundError/ClassNotFoundException in the log "because the bundle wiring for io.etcd.jetcd-all is no longer valid"
+    // can be safely ignored - that's just jetcd/gRPC/Netty not being cleany shut down and still running.. doesn't really matter much.
+  }
 }

--- a/jetcd-osgi/jetcd-karaf/src/test/java/io/etcd/jetcd/osgi/PaxExamWrapperTest.java
+++ b/jetcd-osgi/jetcd-karaf/src/test/java/io/etcd/jetcd/osgi/PaxExamWrapperTest.java
@@ -1,0 +1,61 @@
+/**
+ * Copyright 2017 The jetcd authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.etcd.jetcd.osgi;
+
+import io.etcd.jetcd.launcher.junit.EtcdClusterResource;
+import java.util.Optional;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.JUnitCore;
+import org.junit.runner.notification.Failure;
+
+/**
+ * Pax Exam test wrapper which starts (and stops) an etcd test server via
+ * launcher. We need to do this because jetcd-launcher uses Testcontainers,
+ * which cannot run inside OSGi (due to the use of TCL loadClass(String) in
+ * DockerClientProviderStrategy). And even if we were to use the launcher only
+ * in the config() method (which "is executed before the OSGi container is
+ * launched, so it does run in plain Java"), the Pax Exam probe still needs to
+ * load the entire test class and all of its references (launcher with
+ * Testcontainers) into OSGi, which is a PITA. There is also no easy way to stop
+ * the testcontainer after. It is therefore simplest to just launch the etcd
+ * server before getting Pax Exam's world, like this does.
+ *
+ * @author Michael Vorburger.ch
+ */
+public class PaxExamWrapperTest {
+
+    private static final String ETCD_ENDPOINT_SYSTEM_PROPERTY_NAME = "etcd.endpoint";
+
+    @Rule
+    public final EtcdClusterResource etcd = new EtcdClusterResource("karaf");
+
+    @Test
+    public void testClientServiceChecks() throws Throwable {
+        String endpoint = etcd.cluster().getClientEndpoints().get(0);
+        System.setProperty(ETCD_ENDPOINT_SYSTEM_PROPERTY_NAME, endpoint);
+
+        Optional<Failure> failure = JUnitCore.runClasses(ClientServiceChecks.class).getFailures().stream().findFirst();
+        if (failure.isPresent()) {
+            throw failure.get().getException();
+        }
+    }
+
+    static String getClientEndpoints() {
+        return System.getProperty(ETCD_ENDPOINT_SYSTEM_PROPERTY_NAME);
+    }
+}

--- a/jetcd-osgi/pom.xml
+++ b/jetcd-osgi/pom.xml
@@ -34,7 +34,5 @@
         <module>jetcd-osgi</module>
         <module>jetcd-karaf</module>
     </modules>
-
-
 </project>
 

